packages/model-registry/upstream/frontend/src/odh/utils/deploymentUtils.ts

```diff
@@ -0,0 +1,81 @@
+  const hasModelVersionDeployment = React.useCallback((mvId: string): DeploymentDetectionResult => {
+    if (!loaded || !deployments) {
+      // Conservative approach: assume deployment exists during loading to prevent flicker
+      // This keeps the archive button disabled until we're sure there are no deployments
+      return { hasDeployment: true, loaded };
+    }
+
+    const hasDeployment = deployments.some(deployment => {
+      const isInferenceService = deployment.model?.kind === 'InferenceService';
+      const modelVersionId = deployment.model?.metadata?.labels?.[KnownLabels.MODEL_VERSION_ID];
+      return isInferenceService && modelVersionId === mvId;
+    });
+
+    return { hasDeployment, loaded };
+  }, [deployments, loaded]);
+
+  const hasRegisteredModelDeployment = React.useCallback((rmId: string, modelVersions: ModelVersion[]): DeploymentDetectionResult => {
+    if (!loaded || !deployments) {
+      // Conservative approach: assume deployment exists during loading to prevent flicker
+      return { hasDeployment: true, loaded };
+    }
+
+    // Get all model version IDs for this registered model
+    const modelVersionsForRM = modelVersions.filter(mv => mv.registeredModelId === rmId);
+    const mvIds = modelVersionsForRM.map(mv => mv.id);
+
+    // If no model versions, no deployments possible
+    if (mvIds.length === 0) {
+      return { hasDeployment: false, loaded };
+    }
+
+    // Check if any model version of this registered model is deployed
+    const hasDeployment = deployments.some(deployment => {
+      const isInferenceService = deployment.model?.kind === 'InferenceService';
+      const modelVersionId = deployment.model?.metadata?.labels?.[KnownLabels.MODEL_VERSION_ID];
+      return isInferenceService && modelVersionId && mvIds.includes(modelVersionId);
+    });
+
+    return { hasDeployment, loaded };
+  }, [deployments, loaded]);
+
+  // Alternative method for registered model deployment check when model versions are fetched separately
+  const hasRegisteredModelDeploymentByVersionIds = React.useCallback((mvIds: string[]): DeploymentDetectionResult => {
+    if (!loaded || !deployments) {
+      // Conservative approach: assume deployment exists during loading to prevent flicker
+      return { hasDeployment: true, loaded };
+    }
+
+    if (mvIds.length === 0) {
+      return { hasDeployment: false, loaded };
+    }
+
+    const hasDeployment = deployments.some(deployment => {
+      const isInferenceService = deployment.model?.kind === 'InferenceService';
+      const modelVersionId = deployment.model?.metadata?.labels?.[KnownLabels.MODEL_VERSION_ID];
+      return isInferenceService && modelVersionId && mvIds.includes(modelVersionId);
+    });
+
+    return { hasDeployment, loaded };
+  }, [deployments, loaded]);
+
+  return {
+    hasModelVersionDeployment,
+    hasRegisteredModelDeployment,
+    hasRegisteredModelDeploymentByVersionIds,
+    loaded,
+    deployments
+  };
+};
```

Comment:
If this hook is indeed necessary, I wonder if these two functions `hasRegisteredModelDeployment` and `hasRegisteredModelDeploymentByVersionIds` are redundant. They mostly have the same behavior, maybe we could factor out the duplicated logic and have `hasRegisteredModelDeployment` mostly just get a list of version ids to pass into the current contents of `hasRegisteredModelDeploymentByVersionIds`.

But yeah in general I think we might be filtering by these labels twice. Not sure if I'm missing some piece where we couldn't filter by version id in the context provider or something.